tiledb/sm/compressors/dict_compressor.h

```diff
@@ -0,0 +1,128 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines the dictionary compression class for strings.
+ */
+
+#ifndef TILEDB_DICT_COMPRESSOR_H
+#define TILEDB_DICT_COMPRESSOR_H
+
+#include "tiledb/common/common.h"
+#include "tiledb/sm/misc/endian.h"
+
+#include <map>
+#include <unordered_map>
+#include <vector>
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+/** Handles dictionary compression/decompression of variable sized strings. */
+class DictEncoding {
+ public:
+  /**
+   * Compress variable sized strings into dictionary encoded format
+   *
+   * @tparam T Type of integer that can fit the word ids
+   * @param input Input in form of a memory contiguous sequence of
+   * strings
+   * @param output Dictionary-encoded output in ids of type T. Memory is
+   * allocated and owned by the caller
+   * @return A dictionary in the form of a vector of strings, where indices are
+   * the word_ids of each word
+   */
+  template <class T>
+  static std::vector<std::string_view> compress(
+      const span<std::string_view> input, span<T> output) {
+    if (input.empty() || output.empty()) {
+      return {};
+    }
+
+    // We use string_views in the dictionary, which point to input strings. This
+    // means that input should not be freed before dictionary data is consumed
+    // TBD: Is this acceptable, or do we use dict of strings - more expensive to
+    // create & copy on return
+    std::vector<std::string_view> dict;
+    dict.reserve(input.size());
+    T word_id = 0;
+    std::unordered_map<std::string_view, T> word_ids;
+
+    for (uint64_t i = 0; i < input.size(); i++) {
+      // If we haven't seen that string before, add it to the dictionary
+      if (word_ids.find(input[i]) == word_ids.end()) {
+        dict.emplace_back(input[i]);
+        word_ids[input[i]] = word_id++;
+      }
+
+      output[i] = word_ids[input[i]];
+    }
+
+    // copy happens here, but it's a copy of string_views
+    // TBD: do we want the caller to allocate dict as well?
+    return dict;
+  }
+
+  /**
+   * Decompress strings that are encoded in dictionary format
+   *
+   * @tparam T Type of integer that can fit the word ids
+   * @param input Input dictionary-encoded format of ids of type T
+   * @param dict The dictionary to be used for decoding a word id into a string.
+   * Expected type is a vector of strings, where indices are the ids of
+   * each word
+   * @param output Dictionary-encoded output as a series of ids of size T.
+   * Memory is allocated and owned by the caller
+   */
+  template <class T>
+  static void decompress(
+      const span<T> input,
+      const std::vector<std::string_view>& dict,
```

Comment:
For decompress are we still going to have a vector of string views? This might need to be adjusted when the filter pipeline components are implemented.

Reply:
I don't have a 100% clear idea of how we are going to write `dict` on disk during compression. The `vector` of `string_views` here assumes we will write the vector as a buffer of `chars` with their lengths in the same or another buffer. That way on decompress we can recreate the dict `vector` by constructing `string_views` from that buffer using the lengths and without needing to copy.
Does this address your concern?